http/server.ts

```diff
@@ -1,218 +1,265 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
-import { listen, Conn, toAsyncIterator, Reader, Writer, copy } from "deno";
-import { BufReader, BufState, BufWriter } from "../io/bufio.ts";
+
+import { Conn, copy, listen, Reader, toAsyncIterator, Writer } from "deno";
+import { BufReader, BufWriter } from "../io/bufio.ts";
 import { TextProtoReader } from "../textproto/mod.ts";
 import { STATUS_TEXT } from "./http_status.ts";
 import { assert } from "../testing/mod.ts";
+import { defer, Deferred } from "../async/deferred.ts";
+import { Key, pathToRegexp } from "./path_to_regexp.ts";
```

Comment:
I don't think we should have fancy path matching in this module. Happy to have something like `//http/router.ts` but in this `server.ts` module things should be very simple. 

In Go, ServeMux does not do such complex things: https://golang.org/pkg/net/http/#ServeMux

I would very happily have a ServeMux class here that implemented the above routing logic.

Reply:
@ry dropped path-to-regex, and implemented pure regexp based route matching

Reply:
I'm still concerned that this is using RegExp rather than simple prefix matching... Is it possible to move this routing logic out of server.ts ?

server.ts is meant to be a very low-level web server, which everyone case use. In the case of the routing, I can imagine different people having different solutions.

I think the most complex routing inside server.ts should be limited to what is described here: https://golang.org/pkg/net/http/#ServeMux

Reply:
I don't think that it is much complicated...routing with string or regex can be used for general purpose. `serve()` should be considered as the lowest http server api. 

> Is it possible to move this routing logic out of server.ts ?

Of course. But even if logic become more simple, code won't be changed. We eventually call `req.url.match(pattern)`. Differences are only storing match result in `req` or not. I hesitate that we have two similar server api (ServerMux and Router).